clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -1132,12 +1133,26 @@ private CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> addOffsetFetch
             Optional<OffsetFetchRequestState> inflight =
                     inflightOffsetFetches.stream().filter(r -> r.sameRequest(request)).findAny();
 
-            if (dupe.isPresent() || inflight.isPresent()) {
+            if (dupe.isPresent()) {
                 log.debug("Duplicated unsent offset fetch request found for partitions: {}", request.requestedPartitions);
-                dupe.orElseGet(inflight::get).chainFuture(request.future);
+                dupe.get().chainFuture(request.future);
+            } else if (inflight.isPresent()) {
```

Comment:
I think this area of de-duplicating requests needs a good deal of simplification still. This piece of code in particular is more tricky to understand than is necessary. Here's an example.

* There's a duplicate request in the in-flight requests
* We chain the new request's future onto the existing request
* But the existing request might already been done
* So, then we remove the existing request from the in-flight requests

Does the chaining of futures make the slightest sense in this case?

Reply:
@AndrewJSchofield—I totally agree 😄

The initial pass was my attempt to shoehorn in a fix while disturbing the surrounding code as little as possible. I wanted to keep things as localized as possible, given the deadline.

I am working on an approach that is hopefully more sane.